engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/ContextTest.scala

```diff
@@ -0,0 +1,217 @@
+          |""".stripMargin
+      eval(code) shouldEqual "(Forbidden_Operation.Error 'Input')"
+    }
+
+    "error on invalid environment actions" in {
+      val code =
+        """from Standard.Base import all
+          |from Standard.Base.Runtime.Context import Input
+          |
+          |input_action : Integer -> Integer
+          |input_action i = Input.if_enabled i environment="live"
+          |
+          |main = Panic.catch Any (input_action 2) p-> p.payload.to_text
+          |""".stripMargin
+      eval(
+        code
+      ) shouldEqual "(Not_Implemented.Error execution environment mismatch)"
+    }
+
+    "pass on Input actions" in {
+      val code =
+        """from Standard.Base import all
+          |from Standard.Base.Runtime.Context import Input
+          |
+          |input_action : Integer -> Integer
+          |input_action i = Input.if_enabled i environment="design"
+          |
+          |main = Runtime.with_enabled_context Input (input_action 2)
+          |""".stripMargin
+      eval(code) shouldEqual 2
+    }
+
+    "error on Output actions" in {
+      val code =
+        """from Standard.Base import all
+          |from Standard.Base.Runtime.Context import Output
+          |
+          |output_action : Integer -> Integer
+          |output_action i = Output.if_enabled i environment="design"
+          |
+          |main = Panic.catch Any (output_action 2) p-> p.payload.to_text
+          |""".stripMargin
+      eval(code) shouldEqual "(Forbidden_Operation.Error 'Output')"
+    }
+
+    "scope of context is limited" in {
+      val code =
+        """from Standard.Base import all
+          |from Standard.Base.Runtime.Context import Input
+          |
+          |input_action : Integer -> Integer
+          |input_action i = Input.if_enabled i
+          |
+          |main =
+          |    res = Runtime.with_enabled_context Input (input_action 2)
+          |    Panic.catch Any (input_action 2) p-> res.to_text+" and "+p.payload.to_text
+          |""".stripMargin
+      eval(code) shouldEqual "2 and (Forbidden_Operation.Error 'Input')"
+    }
+
+    "allow locally running IO" in {
+      val code =
+        """from Standard.Base import all
+          |from Standard.Base.Runtime.Context import Input, Output
+          |
+          |output_action : Integer -> Integer
+          |output_action i = Output.if_enabled (i+1) environment="design"
+          |
+          |input_action : Integer -> Integer
+          |input_action i = Input.if_enabled (i*2) environment="design"
+          |
+          |main =
+          |    r = Runtime.with_enabled_context Input environment="design" <| Runtime.with_enabled_context Output environment="design" <| output_action <| input_action 123
+          |    [r].to_text
+          |""".stripMargin
+      eval(code) shouldEqual "[247]"
+    }
+  }
+}
+
+class LiveExecutionEnvironmentTest extends InterpreterTest {
```

Comment:
Two classes in a single file. That's unusual, I'd say.

Reply:
Yeah, it has been the same way with `IOContext` so I just stick to it. I can split it if you want.